src/nvim/lua/vim.lua

```diff
@@ -439,6 +439,22 @@ function vim.notify(msg, log_level, opts) -- luacheck: no unused
   end
 end
 
+local warn_once = (function()
+  local warned = {}
+  return function(msg)
+    if not warned[msg] then
+      vim.api.nvim_echo({{msg, "WarningMsg"}}, true, {})
+      warned[msg] = true
+    end
+  end
+end)()
+
+--- Emit a warning message once per Nvim session.
+---
+---@param msg string Warning message to display to the user.
+function vim.warn_once(msg)
+  warn_once(msg)
+end
```

Comment:
Could use a do block to hide the `warned` from the module's scope.


```suggestion
do
  local warned = {}

  --- Emit a warning message once per Nvim session.
  ---
  ---@param msg string Warning message to display to the user.
  function vim.warn_once(msg)
    if not warned[msg] then
      vim.api.nvim_echo({{msg, "WarningMsg"}}, true, {})
      warned[msg] = true
    end
  end
end
```

Probably personal preference and it is fine for me to keep as is, but I find this would show the intention better than the trick with the ad-hoc function.

Reply:
Yes I agree this is much better, I didn't even think about using a `do` block 👍